esmvaltool/recipes/recipe_preprocessor_test.yml

```diff
@@ -14,6 +14,8 @@ preprocessors:
       # This also works, but allows specifying the pressure coordinate name
       # levels: {dataset: ERA-Interim, coordinate: air_pressure}
       scheme: linear
+    mask_landocean:
+      fx_option: land #if land, will mask land off
```

Comment:
I would simply call this `mask` or maybe `mask_off`.

Reply:
I mean,  `fx_option` --> `mask_out`

Reply:
Before diving into the shapefiles one could try to use land variables to generate land/sea mask if later is not present in the fx files. For example, soil temperature (tsl) is usually masked and land model is often on the same grid as the atmospheric one. You can give the user a choice to try to extract mask from such variables.

Reply:
there is already a function that deals with shapefiles (although I haven't tested it since last year when I put it in, it used to work well) - but I don't think we should complicate ourselves with shapefiles just yet, at least for such a simple task of masking off either land or ocean -- agreed with @koldunovn -- but also recovering a ```tsl``` file means introducing a new variable and a hairy data retrieval function -- I say we wait until the next release for this?

Reply:
Introducing a new variable does not solve the problem, since it could also be not available.

Does Iris offer a simple solution for masking other than using shapefiles @bjlittle ?

Reply:
```maskgeometry()``` and ```mask2d()``` are the functions that are currently in ```_mask.py``` and are masking on lat-lon grids with shapefiles, but that assumes the shapefile exists somewhere -- is there a an iris attribute that can use an already existent land or ocean mask shapefile in iris?

Reply:
As far as I know, shapefiles are freely available for a wide range of purposes, but they are usually pretty large binaries. I don't know whether there is a cheaper solution.

http://openstreetmapdata.com/data/land-polygons

Reply:
aha indeed, but I don't have much experience with them (I only masked Canada out last year :grinning: ) -- the question is how exact are they and what sort of peer review we'd need to be able to use them?

Reply:
There are different level of precision, using them should not be an issue since they are widely accepted as a standard.

Reply:
ok but they're pretty hefty to store them permanently in the ```esmvaltool/``` directory, can you think of a storage location or shall we download them on the fly? They also seem to get updated regularily so we need to keep track of the updates somehow

Reply:
Yes, the storage is an issue, but maybe we can use the low resolution version (23 MB).
It should be more than enough for our purposes.
Updates should not be critical for land/sea masking, I don't expect continental shapes to change that fast.

Reply:
yeah true, if the coastlines change significantly from year to year we'd be pretty srewed :grin: OK, lemme try add these in a temp dir and run with them

Reply:
Instead of adding them to the repo we could let the installation process automatically download them, if possible.

Reply:
one question tho -- what if, out of a few models we are using, only one doesn't have the ```sftlf``` file and we will apply this low res shapefile, albeit the model itself is pretty high res, wouldn't that destroy a whole bunch of data that could influence the model intercomparison or the multimodel stat?

Reply:
sounds good to me, I'll try first with a `fixed location` file, then we can think of how to download them while installing the tool

Reply:
in the meantime @mattiarighi could you pls test with this current implementation when you got some time? I've tested myself but I ran only a couple models from the `_preprocessor_test` namelist

Reply:
Speaking of the shapefiles for boundaries and physical features the [Natural Earth](https://www.naturalearthdata.com/downloads/10m-physical-vectors/10m-ocean/)  is usually the best and simplest choice. All data are public domain. Here are shapefiles for :
[Land](https://www.naturalearthdata.com/downloads/10m-physical-vectors/10m-land/) and [Ocean](https://www.naturalearthdata.com/downloads/10m-physical-vectors/10m-ocean/). Detalisation is good enought for climate models to come in the next 10 years :) and the file size is quite small. 

One can ether make them part of the esmvaltool distribution of download the first time they are called as it is done, for example, in `Cartopy`. The later might be a problem for installations without  internet access.

Another opion is to use [Global Self-consistent, Hierarchical, High-resolution Geography Database (GSHHG)](https://www.ngdc.noaa.gov/mgg/shorelines/gshhs.html). The advantage here might be that you can select level of detalization. Interface to this data is also [availabele in Cartopy](https://scitools.org.uk/cartopy/docs/v0.14/matplotlib/feature_interface.html).

Aftter all, one could use [cartopy interface to download the data](https://scitools.org.uk/cartopy/docs/v0.14/matplotlib/feature_interface.html) since `cartopy` is installed with `iris` by default anyway.

Reply:
Thanks for the detailed overview of all possibilities!
The first options sounds the easiest, we can add the two files to the repo (~6MB) and generate the mask if the `sftlf` variable is nota available for the given model.

Would that work also for irregular grids, also as an alternative to `sftof`?

Reply:
cheers muchly @koldunovn -- come to think of I actually used Natural earth last year when testing the shapefile masking, I'll get those files instead and plug in the functionality, I am currently testing Lee's PR and once done I am onto this

Reply:
@mattiarighi first versione2 now "nota available" -- are you trying to teach us Italian accents? :grinning: :grinning:

Reply:
That's not even italian, I'm just typing too fast :fire:

Reply:
@mattiarighi The disadvantage of having the files stored in the repo is that you don't automatically get the newest version, although I don't think it is critical in our case. 

In the post from @ocefpaf  that you have mentioned above, he creates two dimentional grid from  vectors of `lon` and `lat` with:
```
x, y = np.meshgrid(lon, lat)
```
So I expect the method to work for curvilinear grids and even for unstructured meshes (because it's `ravel`ed afterwards).

One can do something like:
```python
if lon.ndim == 2:
        lon2d, lat2d = lon, lat
elif lon.ndim == 1:
        lon2d, lat2d = np.meshgrid(lon, lat)
``` 
so you cover both rectilinear and curvilinear meshes.

Reply:
hey guys, good news -- I've implemented the method to extract the geometry from the NE file suggested by @koldunovn and to mask with it; bad news -- it is slow as a slow train since it checks for each point from the cube grid if it is in the geometry or not (see my last commit to this branch). Got an idea though -- how about, instead of using shapefiles, we use an existing fx mask from a model that has it and regrid it to the grid the model that needs it has?

Reply:
Looks like the way to speed the things up would be the following:
```python
import shapely.vectorized
mask = shapely.vectorized.contains(region, x_p-180, y_p)
```
This makes it lightning fast and eliminates the need for conversion to `MultiPoint`. You put 2d coordinates in and get directly 2d mask out, pretty amazing :)

Found it here:
https://github.com/Toblerity/Shapely/pull/118

There is one problem with the current aproach, though, that I did not realize - Caspian and Black seas. Sometimes also Mediteranian is not included in the ocean part of the model. The way to proceed might be to add/substract this seas to/from land/ocean.

Reply:
@koldunovn you are legend, man! The solution speeds up the mask building by orders of magnitude :grin: OK, so it's implemented and it works, have a looksee and a test or two @mattiarighi

Reply:
How big are the shapefiles in the repo?

Reply:
```
valeriu@valeriu-PORTEGE-Z30-C:~/ESMValTool_version2_development_alpha$ du -h esmvaltool/preprocessor/ne_masks/
14M	esmvaltool/preprocessor/ne_masks/
```

Reply:
you reckon they too big? @mattiarighi

Reply:
@koldunovn I have just removed the subtraction of 180 degs from `x_p - 180` as it was in your original suggestion, there is no wrapping around so no need to do the reductions, now the maps look correct and there is no Africa on top of America :grin: Also, I can see the Med and the Caspian in the land masking

Reply:
ocen shp reading is slowwww, is there any other ocean file with a lower resolution? maybe this - https://www.naturalearthdata.com/downloads/50m-physical-vectors/ or is it too coarse?

Reply:
@valeriupredoi Glad it worked :) 
`x_p-180` thing was necessary in my case to get the complete map, not just half of it, since `shapely.vectorized.contains` seems to need `[-180 180]` format for lons. Maybe it's version dependent. 

I was trying to say that having Caspian and Med is not nessesary a good thing. Imagine that you have a coupled model with very coarse ocean model, that do not have Mediteranian Sea (this is not uncommon). By creating ocean/land mask, you will exclude/include values from the atmospheric or land models that not supposed to be there, since there is no Med Sea in the Ocean model. Needless to say that ocean models never include Caspian sea (which is actually a lake). I don't think it's extremelly critical, but something to keep in mind :)

I think at present state 1:50m is more than enough. You even can get around with 1:110m for most of the models.

Reply:
@koldunovn  - I actually spent a good bit of today figuring out what the hell was going on with the maps, the NE mask is slap in the middle of the map whereas the regular stflf is wrapped around :grinning: - the last commit addresses this and puts all at the same location on the map, basically one needs to get the left hand of the mask and the right hand of the mask. I tried with the 50m oceans and it is superfast; the 50m land is weird as hell since the shapefile contains in excess of 1400 geometries that I have no clue which one is right; my suggestion: keep land mask at 1:10m and ocean mask at 1:50m?

Reply:
@valeriupredoi I have to apologize, the `x_p-180` is completely wrong, don't know what I was thinking. Hope solution below will simplify the code you have just committed a bit. I still think that one has to convert lons to `[-180 180]` before masking, but `x_p-180` is of course not the proper way to do it. You mask with `x_p_180`, but have to plot with x_p. Sorry once again for the wrong solution :( 
```python
x_p, y_p = np.meshgrid(cube.coord(axis='X').points,
                           cube.coord(axis='Y').points)
x_p_180 = np.where(x_p>=180., x_p-360., x_p)
mask = shapely.vectorized.contains(region, x_p_180, y_p)

import cartopy.crs as ccrs
import matplotlib.pyplot as plt

ax = plt.axes(projection=ccrs.PlateCarree())
ax.coastlines()
plt.contourf(x_p, y_p, cube.data[0,:,:])
plt.show()

ax = plt.axes(projection=ccrs.PlateCarree())
ax.coastlines()
plt.contourf(x_p, y_p, mask)
plt.show()
```